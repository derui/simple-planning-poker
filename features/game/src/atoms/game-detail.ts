import { UseLoginUser } from "@spp/feature-login";
import { Game, GameRepository, User } from "@spp/shared-domain";
import { DeleteGameUseCase } from "@spp/shared-use-case";
import { useAtom, useSetAtom } from "jotai";
import { GameDto, toGameDto } from "./dto.js";
import { gamesAtom, GameStatus, gameStatusAtom, selectedGameAtom } from "./game-atom.js";

/**
 * Hook definition to list game
 */
export type UseGameDetail = () => {
  loading: boolean;

  /**
   * current selected game.
   */
  game?: GameDto;

  /**
   * requesting edit current selected game.
   */
  requestEdit: () => void;

  /**
   * Delete current selected game.
   */
  delete: () => void;
};

/**
 * Create hook implementation of `UseGameDetail`
 */
export const createUseGameDetail = function createUseGameDetail({
  gameRepository,
  useLoginUser,
  deleteGameUseCase,
}: {
  gameRepository: GameRepository.T;
  useLoginUser: UseLoginUser;
  deleteGameUseCase: DeleteGameUseCase;
}): UseGameDetail {
  return () => {
    const [gameStatus, setGameStatus] = useAtom(gameStatusAtom);
    const { userId } = useLoginUser();
    const [game, setGame] = useAtom(selectedGameAtom);
    const setGames = useSetAtom(gamesAtom);
    const loading = gameStatus == GameStatus.Deleting;

    return {
      loading,
      game: game ? toGameDto(game) : undefined,

      requestEdit: (): void => {
        if (!game || !userId) {
          return;
        }

        setGameStatus(GameStatus.Edit);
      },

      delete: () => {
        if (!game || !userId || loading) {
          return;
        }

        const _do = async () => {
          try {
            const ret = await deleteGameUseCase({ gameId: Game.createId(game.id), ownedBy: User.createId(userId) });

            if (ret.kind == "success") {
              setGame(undefined);
              setGames(await gameRepository.listUserCreated(User.createId(userId)));
            }
          } catch (err) {
            console.error(err);
          } finally {
            setGameStatus(GameStatus.Detail);
          }
        };

        setGameStatus(GameStatus.Deleting);
        _do();
      },
    };
  };
};
