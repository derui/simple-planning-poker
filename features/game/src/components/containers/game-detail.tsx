import { useCallback } from "react";
import { GameDetailLayout } from "./game-detail.layout.js";

import { useCurrentGame } from "../../atoms/use-current-game.js";
import { useEditGame } from "../../atoms/use-edit-game.js";

interface Props {
  /**
   *  Notify the parent that the game has been voted.
   */
  onStartVoting?: (gameId: string) => void;
}

export const GameDetail = function GameDetail({ onStartVoting }: Props): JSX.Element {
  const { loading, delete: _delete, game } = useCurrentGame();
  const { edit, loading: editing } = useEditGame();

  const handleDelete = useCallback(() => {
    _delete();
  }, [_delete]);

  const handleStartVoting = useCallback(() => {
    if (game) {
      onStartVoting?.(game?.id);
    }
  }, [game, onStartVoting]);

  const handleSubmit = useCallback(
    (name: string, points: string) => {
      edit(name, points);
    },
    [edit]
  );

  return (
    <GameDetailLayout
      game={game}
      onDelete={handleDelete}
      onSubmit={handleSubmit}
      onStartVoting={handleStartVoting}
      loading={loading || editing}
    />
  );
};
