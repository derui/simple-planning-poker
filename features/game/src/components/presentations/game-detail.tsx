import { Variant } from "@spp/shared-color-variant";
import { Icon } from "@spp/ui-icon";
import * as styles from "./game-detail.css.js";

export interface Props {
  readonly name?: string;

  readonly points?: string;

  /**
   * The callback to call when the user clicks on the edit button.
   */
  readonly onEdit?: () => void;

  /**
   * The callback to call when the user want to delete this game.
   */
  readonly onDelete?: () => void;

  /**
   * The callback to call when the user wants to start voting.
   */
  readonly onStartVoting?: () => void;
}

export const GameDetail = function GameDetail({ name, points, onEdit, onDelete, onStartVoting }: Props): JSX.Element {
  if (!name || !points) {
    return (
      <div className={styles.root}>
        <div className={styles.empty}>Select game from list</div>
      </div>
    );
  }

  return (
    <div className={styles.root}>
      <button className={styles.editButton} onClick={onEdit} aria-label="Edit">
        <Icon.Pencil variant={Variant.emerald} />
      </button>
      <div className={styles.defList}>
        <span className={styles.defListLabel}>Name</span>
        <span className={styles.defListContent}>{name}</span>
      </div>
      <div className={styles.defList}>
        <span className={styles.defListLabel}>Points</span>
        <span className={styles.defListContent}>{points}</span>
      </div>
      <div className={styles.footer}>
        <button className={styles.deleteButton} onClick={onDelete}>
          <Icon.Trash variant={Variant.cerise} />
          Delete
        </button>
        <button className={styles.startVotingButton} onClick={onStartVoting}>
          <Icon.Check variant={Variant.emerald} />
          Start Voting
        </button>
      </div>
    </div>
  );
};
