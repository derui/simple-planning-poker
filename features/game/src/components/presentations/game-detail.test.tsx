import { cleanup, render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import Sinon from "sinon";
import { afterEach, expect, test } from "vitest";
import { GameDetail } from "./game-detail.js";

afterEach(cleanup);

test("should be able to render", async () => {
  const ret = render(<GameDetail name="name" points="points" />);

  expect(ret.container).toMatchSnapshot();
});

test("should call callback onEdit when edit button is clicked", async () => {
  // Arrange
  const mockCallback = Sinon.fake();
  render(<GameDetail name="name" points="points" onEdit={mockCallback} />);

  // Act
  await userEvent.click(screen.getByLabelText("Edit"));

  // Assert
  expect(mockCallback.calledOnce).toBeTruthy();
});

test("should call callback onDelete when delete button is clicked", async () => {
  // Arrange
  const mockCallback = Sinon.fake();
  render(<GameDetail name="name" points="points" onDelete={mockCallback} />);

  // Act
  await userEvent.click(screen.getByRole("button", { name: "Delete" }));

  // Assert
  expect(mockCallback.calledOnce).toBeTruthy();
});

test("should call callback onStartVoting where start voting button is clicked", async () => {
  // Arrange
  const mockCallback = Sinon.fake();
  render(<GameDetail name="name" points="points" onStartVoting={mockCallback} />);

  // Act
  await userEvent.click(screen.getByRole("button", { name: "Start Voting" }));

  // Assert
  expect(mockCallback.calledOnce).toBeTruthy();
});

test("should show empty if name is not passed", () => {
  // Arrange
  render(<GameDetail />);

  // Act

  // Assert
  expect(screen.queryByText(/Select game/)).not.toBeNull();
});
