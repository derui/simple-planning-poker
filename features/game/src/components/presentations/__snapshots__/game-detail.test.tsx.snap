// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be able to render 1`] = `
<div>
  <div
    class="game-detail_root__6k3kyt0"
  >
    <button
      aria-label="Edit"
      class="game-detail_editButton__6k3kyt2 style_enabledIconButtonVariants_emerald__92u52z10 style_enabledIconBase__92u52z5 style_iconBase__92u52z1 style_transition_all__192bcwg41"
    >
      <span
        class="pencil_medium_emerald__2ixuvsc style_iconMedium__p90ga82 style_iconBase__p90ga80"
      />
    </button>
    <div
      class="game-detail_defList__6k3kyt3"
    >
      <span
        class="game-detail_defListLabel__6k3kyt4"
      >
        Name
      </span>
      <span
        class="game-detail_defListContent__6k3kyt5"
      >
        name
      </span>
    </div>
    <div
      class="game-detail_defList__6k3kyt3"
    >
      <span
        class="game-detail_defListLabel__6k3kyt4"
      >
        Points
      </span>
      <span
        class="game-detail_defListContent__6k3kyt5"
      >
        points
      </span>
    </div>
    <div
      class="game-detail_footer__6k3kyt6"
    >
      <button
        class="game-detail_deleteButton__6k3kyt7 style_enabledButtonVariants_cerise__92u52zu style_enabledBase__92u52z3 style_base__92u52z0 style_transition_all__192bcwg41"
      >
        <span
          class="trash_medium_cerise__1a232vqf style_iconMedium__p90ga82 style_iconBase__p90ga80"
        />
        Delete
      </button>
      <button
        class="game-detail_startVotingButton__6k3kyt8 style_enabledButtonVariants_emerald__92u52zr style_enabledBase__92u52z3 style_base__92u52z0 style_transition_all__192bcwg41"
      >
        <span
          class="check_medium_emerald__1oavie2c style_iconMedium__p90ga82 style_iconBase__p90ga80"
        />
        Start Voting
      </button>
    </div>
  </div>
</div>
`;
